Post processing shader system overhaul!

iwubcode opened the pull request:
TODO:

- [ ] Fix OpenGL (somehow the last few weeks it became completely broken)
- [ ] Support D3D12 multi output for compute shaders
- [ ] Fix broken depth (screenshot is from old build)
- [ ] Fix error reporting in UI
- [ ] Allow saving profiles for shaders
- [ ] Code cleanup

Have you ever cried into your pillow at night wishing that Dolphin used *more* of your GPU?  Well hang on to your star bits because after a year (off and on) of development, a new system is here that might help you get your wish.  That's right, the revamped post processing system is now up for review!

**Note for reviewers:** When I set out to write the new system, I didn't want to do things a little better.  I didn't want to just match other software out there in terms of functionality.  I wanted to punch-out the competition and set a new example for what you could do with enhancements in an emulator.  So my apologies in advance for the large amounts of code, I am happy to split out into separate PRs.

## Features

### Configuration

The new system uses json for configuration to handle the structure needed to support all the new features.

```
{
	"meta":
	{
		"author": "iwubcode, Fubaxiusz",
		"description": "A shader that draws a color LUT (lookup-table)"
	},
	"options":
	[
		{
			"type": "int",
			"name": "_LutSize",
			"ui_name": "LUT Block Size",
			"min": [ 16.0 ],
			"max": [ 64.0 ],
			"default": [ 32.0 ]
		},
		{
			"type": "bool",
			"name": "LUT_VERTICAL",
			"ui_name": "LUT Vertical",
			"default": false,
			"is_constant": true
		}
	],
	"passes":
	[
		{
			"entry_point": "main",
			"inputs":
			[
				{
					"type": "previous_pass",
					"texture_filter": "linear",
					"texture_mode": "clamp"
				}
			]
		}
	]
}
```

### Multiple Shaders

The first feature you may notice is that you can now have multiple shaders in one go.  How you order the shaders affects the outcome:

![shaders](https://user-images.githubusercontent.com/15224722/148655364-bc706c31-e50c-499b-a861-df5e4af44c15.gif)

### Multiple Passes

Shaders can now use multiple passes.  Each pass specifies a different entry point in the shader.

### Depth

The biggest feature users ask for when they talk about shaders is leveraging depth effects and that is possible:

![00000000_2021-12-04_00-09-38](https://user-images.githubusercontent.com/15224722/148655436-1dc430ff-cd00-4ac5-8d44-cc831d3824f7.png)

### Compute Shaders

The code supports general purpose compute shaders, here's an emboss effect done with compute:

![tos_compute](https://user-images.githubusercontent.com/15224722/148655525-f6af2d1d-decd-4858-9ee8-aa49c5e8e072.png)

### Multiple Output

Shaders can be configured to output to multiple output textures.  Output can then be accessed later on by name.  This can be more performant than using multiple passes and makes porting from other systems easier.

### Upscaling / Downscaling

Shaders internally support an "output scale" which allows the resulting image to be upscaled or downscaled.


## UI

### New controls

New controls have been added to support color selection or a list of options via a drop-down.

![outline_shader](https://user-images.githubusercontent.com/15224722/148657161-ffb2a7f0-431e-4661-ad90-5a83c53bc2ca.png)

### Images

Images can be defined by the shader, these images are not exposed to the user.

The shader can also define user images.  This allows the user to modify them using a button to select the image.  This control also supports drag and drop:

![image_drag_drop](https://user-images.githubusercontent.com/15224722/148657106-2a6d5f6d-6636-4fc5-8e11-250004f0e9a5.gif)

### Grouped controls

Controls can be in a group, this puts them under a tab:

![lightroom-shader](https://user-images.githubusercontent.com/15224722/148655936-57e83b9b-9a9d-448c-81ca-acbd633e9d07.gif)

### Lightroom inspired snapshots

You can load and save snapshots of options, so as to try out different presets to find what you are happy with.

![snapshots](https://user-images.githubusercontent.com/15224722/148661767-15b931d2-ead5-48ea-9aec-95bc35b7fcfb.gif)

### Tooltips

Tooltips can be defined for shader options:

![tooltip](https://user-images.githubusercontent.com/15224722/148657037-ac6f57cb-7897-4c88-8711-d074d23b2ffe.png)

### Adding Shaders

When adding a shader, you can pick from a system shader (which is categorized by 'type'):

![shader-config-window-add](https://user-images.githubusercontent.com/15224722/148655959-8acc5ed7-98c3-4124-a360-488cfdc6f3ac.png)

Or a user shader (with drag/drop capability).

### Stereoscopic changes

Anaglyph and passive stereoscopic modes behaved strangely with the old system.  Using one would set a shader in the post processing system, meaning you couldn't use any other shader.  The new system applies anaglpyh or passive stereoscopic shaders *after* the general post processing shaders to ensure you can avoid your very own disaster, day of crisis.

![anaglyph](https://user-images.githubusercontent.com/15224722/148658432-20c5d6b8-6d41-485f-a81e-807cd5b1017c.png)

## Graphics Triggers

In particular for depth effects, there's no generic place to trigger effects for all games.  I tried a number of different solutions and in the end I decided a game-specific system is needed.  In order to trigger based on different criteria, , I came up with the idea of the "graphics trigger".  This is a system allowing you to define various "profiles" for graphics effects.  Example defining a trigger for when a game applies bloom.

The idea is this would be used for more than just post processing but also other graphics features like [hud moving / bloom removal](https://github.com/dolphin-emu/dolphin/pull/9779).

Here is an example of the "EFB" trigger dialog:

![efb_graphics_trigger](https://user-images.githubusercontent.com/15224722/148657258-9063a155-7594-4f6e-9aa7-c00e3a44b05c.png)

System and user triggers can be defined.

For post processing, I added system triggers that matched the original PR ( #3366 )

Miksel12 commented:
Awesome to see this come to fruition! I did some brief testing and it looks awesome!
There was a crash when closing the game and D3D12 didn't like the shaders but that is to be expected from a wip pr.
Something that can be improved in the UI I think, is the trigger selection dropdown. It is currently located in a sub menu which makes it a bit of a chore to change. I don't see any reason to not place the dropdown on the main screen.

iwubcode commented:
Thank you for testing @Miksel12 !

> There was a crash when closing the game and D3D12 didn't like the shaders but that is to be expected from a wip pr.

Crash on what backend?  Also, what do you mean by D3D12 didn't like the shaders?  Which one?

> Something that can be improved in the UI I think, is the trigger selection dropdown. It is currently located in a sub menu which makes it a bit of a chore to change. I don't see any reason to not place the dropdown on the main screen.

Yeah, if I remember right, I think part of the reason might be for reuse but I can likely make it a combo box class that can be reused.  Only advantage a separate dialog would provide is if we wanted to do something more advanced but I don't have any ideas at the moment.

BhaaLseN commented:
Thoughts on adding a JSON schema for those descriptor files? See `docs/game-mod-descriptor.json` from #10187 and #10334 for an example.

Nothing more annoying than writing a file that looks correct but missing a level, letter or having some other typo that makes it invalid.

Miksel12 commented:
> Crash on what backend? Also, what do you mean by D3D12 didn't like the shaders? Which one?

D3D12 gives the error: "Create failed in ...VideoBackends\D3D12\DX12Texture.cpp at line 119: Create D3D12 texture resource" multiple times, after clicking ok on all of them, Dolphin crashes. This happens without any post processing enabled. W10, GTX 1060 on 497.09

iwubcode commented:
> D3D12 gives the error: "Create failed in ...VideoBackends\D3D12\DX12Texture.cpp at line 119: Create D3D12 texture resource" multiple times, after clicking ok on all of them, Dolphin crashes. This happens without any post processing enabled. W10, GTX 1060 on 497.09

Do you use D3D12 normally without issue?  I am not able to reproduce that error.

>  Thoughts on adding a JSON schema for those descriptor files?

I saw that and am not opposed to doing so but do we even have support for json validation?

BhaaLseN commented:
> > Thoughts on adding a JSON schema for those descriptor files?
> 
> I saw that and am not opposed to doing so but do we even have support for json validation?

Main aspect is that the ones creating them get syntax completion and stuff, to help them get it right the first time. I don't _think_ we validate the files as such (_yet_).

Validating them, or at least returning errors while reading is a plus though, but that can always come later to limit the scope of this PR (just like the schema itself; was mainly to bring it up, not necessarily to have it implemented in here if you want to defer it for later).

iwubcode commented:
Makes sense.  My plan was to add some documentation, though I wasn't sure if I'd do it as a subsequent PR as this one is already insanely large (assuming it isn't broken apart).

Miksel12 commented:
> Do you use D3D12 normally without issue? I am not able to reproduce that error.

Yes, D3D12 on master works without any problems.

JMC47 commented:
I'm stupid and dumb, so I don't really know how to make shaders, but in my use this is pretty cool.

iwubcode commented:
Hmm, in my tests OpenGL is now fixed (I had to clear my shader cache).  Strange that buildbot says it isn't.  Could someone else check?

I also fixed the shutdown crash.

AaronBPaden commented:
Trying to see if I can port MAME's crt-geom-deluxe over using this new system. I don't really understand shaders but I can read code well enough to do a port job. Well unless it's too hard and I give up. 50/50 chances heh

But does this new system support vertex shaders? I think all the included shaders are fragment shaders, which is a problem I ran into last time I tried something similar with the old system.

iwubcode commented:
> Trying to see if I can port MAME's crt-geom-deluxe over using this new system. I don't really understand shaders but I can read code well enough to do a port job. Well unless it's too hard and I give up. 50/50 chances heh

Just full disclosure, I may remove or rename some of the shader functions that are currently exposed.  I'm still thinking about it.

> But does this new system support vertex shaders? I think all the included shaders are fragment shaders, which is a problem I ran into last time I tried something similar with the old system.

It does not.  A post process is done with a single quad (4 vertices).  I take the same stance as the previous author who attempted to implement this, I don't really see any use cases where we need a vertex shader.  Any effect should be doable with a fragment shader or compute shader.

If we _really_ need vertex shader support for some use case, we can look into adding it at a later date.

AaronBPaden commented:
> Just full disclosure, I may remove or rename some of the shader functions that are currently exposed. I'm still thinking about the final "api".

OK, thanks for the heads up!

> It does not. A post process is done with a single quad (4 vertices). I take the same stance as the previous author who attempted to implement this, I don't really see any use cases where we need a vertex shader. Any effect should be doable with a fragment shader or compute shader.

Oh yeah, maybe. Well like I said, my shader knowledge is limited. I would have **thought** that the geometry warping part would be done with a vertex shader, but I'm looking at it now in crt-geom-deluxe and it looks like it just does the standard thing and is just using the vertex shader to calculate some values that are sent to the fragment shader anyway.

So maybe it isn't necessary, as long as the fragment shader has access to enough information. But it would definitely make porting the other shaders that are around more straightforward.

Actually, looking at some files I have lying around I think I came to the conclusion before that I could just put everything in the fragment shader and the real issue I ended up having was with trying to combine multiple shaders into one file. This led to all kinds of fun and ultimately me just using a bilinear shader for Dolphin hehe.

Multiple passes should make this more manageable. I'm interested in seeing what I can do. I'll try and put together a demo.

iwubcode commented:
> Actually, looking at some files I have lying around I think I came to the conclusion before that I could just put everything in the fragment shader and the real issue I ended up having was with trying to combine multiple shaders into one file. This led to all kinds of fun and ultimately me just using a bilinear shader for Dolphin hehe.

Yes, there could be some more work to be done to share shader functions amongst multiple shaders.  When porting a number of shaders, I didn't really need that but I could see how that would be beneficial.  That may be something I look into in the future.

> Multiple passes should make this more manageable. I'm interested in seeing what I can do. I'll try and put together a demo.

Looking forward to it.  Let me know if you hit any major road blocks.

AaronBPaden commented:
OK, I have a couple of quick ports done which I've published [here](https://github.com/AaronBPaden/dolphin-shader-demo).

I went ahead and ported CRT Lottes because it's simpler. It doesn't look **great**. If the emulator outputs to 480p you really need to pair your CRT shader with a shader that simulates interlacing I think to get good results. But it looks alright and it'll do for now for a demo.

I've did also port the lowpass filter from crt-geom-deluxe to give it a slight analog blur.

A couple problems I noticed:

* The warping feature of CRT Lottes doesn't work here. If you change warpX and warpY you'll see that the edges become smudged, not rounded. Maybe you have some insight here?
* Weirdly-ranged inputs seem to be broken in the options UI. The options in CRT Lottes that started negative always started at 0.0 in the UI. The lowpass settings which have large numbers also start at 0 and I can't change them at all in the UI.

A couple screenshots showing off (needs to be viewed full screen to see the scanline effect):

![2022-02-06-20:26:51](https://user-images.githubusercontent.com/1188012/152717257-0198b6b6-872e-4fb0-a1a5-b3779ac0663b.png)

![2022-02-06-20:31:20](https://user-images.githubusercontent.com/1188012/152717263-654eefc1-07bf-41db-9c0a-119073eae89a.png)

iwubcode commented:
> I went ahead and ported CRT Lottes because it's simpler. It doesn't look great. If the emulator outputs to 480p you really need to pair your CRT shader with a shader that simulates interlacing I think to get good results. But it looks alright and it'll do for now for a demo.
I've did also port the lowpass filter from crt-geom-deluxe to give it a slight analog blur.

Great job on the demo.  I ran your crt shader and had to make some changes because it didn't run on D3D.  A couple minor comments for portability:

* Use float3 instead of vec3, float2 instead of vec2, etc
* HLSL doesn't like it when you do `float2(0.5)`.  If you intend to do float2, do `float2(0.5, 0.5)`.  There was one place I needed to introduce a variable because of that.

> The warping feature of CRT Lottes doesn't work here. If you change warpX and warpY you'll see that the edges become smudged, not rounded. Maybe you have some insight here?

Not sure where this was ported from but you're dealing with textures on a quad outputted to the screen.  Because of that, you will always have a square picture with distribution of the pixel data.  Are you expecting black or some other color on the edges to make it look spherical?  If so, you'd have to output that color at that offset.

I guess this was ported from some vertex shader.  In that scenario, I guess a vertex shader would be useful, as you could actually bend the vertices into that form (and the background color would show behind).  We don't have that luxury.

> Weirdly-ranged inputs seem to be broken in the options UI. The options in CRT Lottes that started negative always started at 0.0 in the UI. The lowpass settings which have large numbers also start at 0 and I can't change them at all in the UI.

Thank for reporting that!  I fixed that issue.

AaronBPaden commented:
Thanks! I can't test on D3D, but I tried to fix those issues.

CRT Lottes actually doesn't use the vertex shader for anything. But the warping would never have worked, because I botched one line! I think I meant to come back to it and forgot.

Anyway, I've updated my repo with what I have so far, but I'm a little confused about the API. Here's the line in question, from my currently-working dosbox-staging version of the shader:

```
vec2 pos = Warp(v_texCoord.xy*(rubyTextureSize.xy/rubyInputSize.xy))*(rubyInputSize.xy/rubyTextureSize.xy);
```

Here's what it looks like in Dolphin now that I've updated it:

```
float2 pos = Warp(GetCoordinates()*(GetResolution()/GetSrcResolution()))*(GetSrcResolution()/GetResolution());
```

Warp, and its associated functions don't use any any API, so they're unchanged. I **think** I've used the right functions here, but I find the naming conventions in both a little confusing. I actually have some ideas about the API related to that.

I believe `rubyInputSize` and `GetSrcResolution` refer to the actual emulated resolution. So `[640, 528]`. I think a new, more explicit API might be called `GetInternalResolution`, which is a concept any Dolphin user would be familiar with. I **assume** `GetResolution()` and `rubyTextureSize` refer to the same thing, but I wasn't sure. A `GetTextureResolution()` might help here. Also `GetViewportResolution()` might make more sense than `GetWindowResolution()`, though that's nit-picky.

Anyway, It still doesn't quite work, and I may be using the wrong functions. But I do also think something *may* be something going on external to the shader, because the warping looks different in different scenes. In Super Mario Sunshine, the intro cinematic does look like the display is being curved, while in gameplay the texture is completely square and the edges are simply smudged. Could also be because of differences between OpenGL and Vulkan, IDK.

--

Oh, I have one more issue to report. I've discovered sometimes Dolphin does not remember what order shaders are loaded in after closing and re-opening. I'm not 100% what the trigger is, if it's random or what. But right now dolphin-emu is insisting on loading crt-lottes before lowpass on startup. Other times lowpass has loaded before crt-lottes.

iwubcode commented:
> CRT Lottes actually doesn't use the vertex shader for anything. But the warping would never have worked, because I botched one line!

Ah I'll have to retry with your fixed version.

> I actually have some ideas about the API related to that

A lot of what you're mentioning is the old API.  I have thought and am still considering revamping or tossing some of it.  I'm used to seeing it now, so I can't say whether it's confusing but regardless of which approach I go with, it needs some documentation.

> But I do also think something may be something going on external to the shader, because the warping looks different in different scenes

I'll have to look at the logic again.  I certainly wouldn't expect any differences between OGL and Vulkan.

> Oh, I have one more issue to report. I've discovered sometimes Dolphin does not remember what order shaders are loaded in after closing and re-opening. I'm not 100% what the trigger is, if it's random or what

This is definitely surprising.  The way the logic works is we save a json file into the Dolphin config directory at shutdown.  This file contains the shaders and the ordering.  In my testing, I haven't noticed the bug you speak of but I'll keep an eye out for it.

> Almost forgot! I also wanted to say that depending on the number of options the configuration window can easily extend past the maximum height of a 1080p monitor. Somebody with a 1366x768 laptop would definitely complain eventually at some point as well. :) I'd probably solve that by constraining the height of the window and adding scrollbars to the shader and options lists.

Yes this is a good idea.  Until I add a scrollbar, I'd suggest using groups.

TryTwo commented:
For shaders, can you make it so only one option can be checked in a group of options? Maybe have one option enable/disable other options as well.

iwubcode commented:
@TryTwo 

> can you make it so only one option can be checked in a group of options?

You should be able to just use a combobox for that?

> Maybe have one option enable/disable other options as well.

This was something I had in my first draft, I can look into adding it back.

TryTwo commented:
> You should be able to just use a combobox for that?

Oh right, you can't currently do a combobox, so I didn't think of that.

> This was something I had in my first draft, I can look into adding it back.

I guess the current PP has a dependent option you can include, but it's linked to tabs and stuff.

TryTwo commented:
How do you feel about adding a bump mapping shader?  Years ago I tried to port some shaders to dolphin, not sure if any of that work would be of interest to you. [My thread](https://forums.dolphin-emu.org/Thread-pack-of-ported-shaders)

also: [DolphinFX shaders](https://forums.dolphin-emu.org/Thread-dolphinfx-post-processing-suite-for-the-opengl-backend)

TryTwo commented:
How do you feel about adding a bump mapping shader?  Years ago I tried to port some shaders to dolphin, not sure if any of that work would be of interest to you. [My thread](https://forums.dolphin-emu.org/Thread-pack-of-ported-shaders)

also: [DolphinFX shaders](https://forums.dolphin-emu.org/Thread-dolphinfx-post-processing-suite-for-the-opengl-backend)

iwubcode commented:
Post processing graphics mods are now possible!  A longer explanation at the bottom of this post.

In addition, I fixed a bug in graphics mods where the draw-started wasn't working for EFBs.  This made it impossible to skip the EFB draws which for Arc Rise Fantasia left a nice box in the corner...

No Graphics Mods:
![arc-no-mods](https://user-images.githubusercontent.com/15224722/180149329-3b33b59e-55e3-4625-bf62-4e07096e6668.png)


Graphics Mods (Skip Bloom, Skip "fixup" bloom copy game does, custom 'pirate bloom' ported from Reshade):
![arc-with-mods](https://user-images.githubusercontent.com/15224722/180149491-dba9833c-ec10-4a63-b37d-2d28d2b8e229.png)

----------------------------------------

This change was rather big in that it _removed_ the user focus of the post processing system.  Users can still apply normal post processing shaders (color correction, custom bloom, etc) all on top of the final outputted image like Dolphin master.  However, if they want to do anything fancy (depth, triggering before the HUD or something else), they need a graphics mod.

I ended up going this route because the UI for post processing would be really convoluted if I tried to account for different graphics mod targets.  That and, making these changes is difficult.  This isn't like Reshade where you choose an effect and poof 95% of your games work magically.  Depth effects or effects triggered at certain times can be a pain (see Arc Rise above).

In the future, I hope to have an editor to make graphics mods easier to create.  At the moment, the mod above was built like this:

* Took the pirate bloom user shader I had.  I saved a profile for it
* Copied that profile and the shader+its json into the graphics mod directory
* Updated the profile to change the "user" source to "mod" and fixed up the path
* Added the profile to the graphics mod metadata

Here's the metadata file:

```
{
	"meta":
	{
		"title": "Pirate Bloom",
		"author": "iwubcode",
		"description": "Ported reshade bloom shader 'Pirate Bloom'"
	},
	"features":
	[
		{
			"group": "Bloom",
			"action": "post_process",
			"action_data": {
				"profile_name": "profile.ini"
			}
		}
	]
}
```

As you can see, not exactly user friendly right now.

TryTwo commented:
Seeing this mentioned again, the UI stuff is my most expected feature, as it opens up a whole lot to the end user.  I would love a bit of an update on the progress and goals.  I don't think it needs to be perfected, the stuff shown here years ago is enough to warrant its use.

iwubcode commented:
>  I would love a bit of an update on the progress and goals

Unfortunately, not much update right now.  The main focus has been getting the custom shader solution cleaned up, graphics mods 2.0, as well as the graphics mod editor.  Some of those features will impact post processing, which is why I stopped working on it in the first place.

Luckily, most of the difficult part of this PR (the graphics backend changes to Dolphin) have been merged to master in the last couple years.  So there's really the post processing logic itself and the UI left.

Version 3 of post processing has the following tasks:

1. Separate backends for different post processing sources.  We'll still have Dolphin but we might support Reshade's API for instance.  I haven't done much research here so it really depends on whether this is a good fit but I'd like to support other systems if possible.
2. Initial implementation will be built against graphics mods and integrated into the editor using imgui.  This system will be more geared toward modders and likely have more features available (named render targets from custom shaders, depth access, etc).  Modders will have the ability to target when the post processing shader gets triggered (ex: replacement bloom for normal game bloom)
3. Something similar to what is shown here will be done for users.  Most likely it will be in qt.  To reduce complexity, this will only allow post processing of the final color output (ex: color correction, scanline shaders, etc).  This should mainly be UI work and some configuration, all the post processing functionality should be done by then.  I hope this will largely be lift and shift from this implementation.

One outstanding feature I'd like to add to custom shaders that would be nice for post processing as well, is a library/include system.  Not sure if I will tackle this first or wait until I'm further along.

My hope is to start on post processing v3 sometime this summer (late July or August) but things tend to take longer than I expect 😄 .
